File: Java/ArrayListAndStreamPractice/src/Anagrams/AnagramValidation.java
package Anagrams;

import java.util.Arrays;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

public class AnagramValidation {

    String string1;

    String string2;

    private Map<Character, Long> getFrequencyCount(String string) {
        return string.chars()
                .mapToObj(x -> (char) x)
                .collect(Collectors.groupingBy(x -> x, Collectors.counting()));
    }


    public AnagramValidation(String string1, String string2) {
        this.string1 = string1;
        this.string2 = string2;
    }

    private String sortStringAlphabetically(String string) {
        char[] charArray = string.toCharArray();
        Arrays.sort(charArray);
        return String.valueOf(charArray);
    }

    public boolean validateByMeansOfStringComparison() {
        return Objects.equals(sortStringAlphabetically(string1), sortStringAlphabetically(string2));
    }

    public boolean validateByMeansOfMapComparison() {
        return Objects.equals(getFrequencyCount(string1), getFrequencyCount(string2));
    }
}

File: Kotlin/higherOrderArrayMethods/src/anagrams/AnagramValidation.kt
package anagrams

class AnagramValidation(
    private val string1: String,
    private val string2: String
) {

    private fun getFrequencyCount(string: String): MutableMap<Char, Int> {
        val stringAsArray = string.split("")
        val frequencyCount = mutableMapOf<Char, Int>()
        for (char in stringAsArray.subList(1, stringAsArray.size - 1).map{ i -> i.single()}) {
            if(frequencyCount[char] != null) {
                frequencyCount[char] = frequencyCount[char]!! + 1
            } else {
                frequencyCount[char] = 1
            }
        }
        return frequencyCount
    }
    private fun sortStringAlphabetically(string: String): String {
        val stringAsArray = string.split("")
        return stringAsArray.subList(1, stringAsArray.size - 1).sorted().toString()
    }
    fun validateByMeansOfStringComparison(): Boolean {
        return sortStringAlphabetically(string1) == sortStringAlphabetically(string2)
    }

    fun validateByMeansOfMapComparison(): Boolean {
        return getFrequencyCount(string1) == getFrequencyCount(string2)
    }


}